packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js

```diff
@@ -1258,6 +1402,97 @@ QUnit.module('Menu tests', {
+
+        const nestedSubmenu1 = $subItem1.find(`.${DX_SUBMENU_CLASS}`)[0];
+
+        assert.strictEqual(getDomNode(handlerShowing.args[1][0].rootItem), $rootItem[0], '2nd level show');
+        assert.strictEqual(handlerShowing.args[1][0].itemData.id, 11);
+        assert.strictEqual(getDomNode(handlerShowing.args[1][0].submenuContainer), nestedSubmenu1);
+        assert.strictEqual(getDomNode(handlerShown.args[1][0].rootItem), $rootItem[0]);
+        assert.strictEqual(handlerShown.args[1][0].itemData.id, 11);
+        assert.strictEqual(getDomNode(handlerShown.args[1][0].submenuContainer), nestedSubmenu1);
+
+        $($subItem2).trigger('dxclick');
+
+        const nestedSubmenu2 = $subItem2.find(`.${DX_SUBMENU_CLASS}`)[0];
+
+        assert.strictEqual(getDomNode(handlerHiding.args[0][0].rootItem), $rootItem[0], '2nd level hide');
+        assert.strictEqual(handlerHiding.args[0][0].itemData.id, 11);
+        assert.strictEqual(getDomNode(handlerHiding.args[0][0].submenuContainer), nestedSubmenu1);
+        assert.strictEqual(getDomNode(handlerHidden.args[0][0].rootItem), $rootItem[0]);
+        assert.strictEqual(handlerHidden.args[0][0].itemData.id, 11);
+        assert.strictEqual(getDomNode(handlerHidden.args[0][0].submenuContainer), nestedSubmenu1);
+
+        assert.strictEqual(getDomNode(handlerShowing.args[2][0].rootItem), $rootItem[0], 'open 2nd level again');
+        assert.strictEqual(handlerShowing.args[2][0].itemData.id, 12);
+        assert.strictEqual(getDomNode(handlerShowing.args[2][0].submenuContainer), nestedSubmenu2);
+        assert.strictEqual(getDomNode(handlerShown.args[2][0].rootItem), $rootItem[0]);
+        assert.strictEqual(handlerShown.args[2][0].itemData.id, 12);
+        assert.strictEqual(getDomNode(handlerShown.args[2][0].submenuContainer), nestedSubmenu2);
+
+        $(document).trigger('dxpointerdown');
+
+        assert.strictEqual(getDomNode(handlerHiding.args[1][0].rootItem), $rootItem[0], '1st level hide');
+        assert.strictEqual(handlerHiding.args[1][0].itemData.id, 1);
+        assert.strictEqual(getDomNode(handlerHiding.args[1][0].submenuContainer), rootSubmenu);
+        assert.strictEqual(getDomNode(handlerHidden.args[2][0].rootItem), $rootItem[0]);
+        assert.strictEqual(handlerHidden.args[2][0].itemData.id, 1);
+        assert.strictEqual(getDomNode(handlerHidden.args[2][0].submenuContainer), rootSubmenu);
+
+        assert.strictEqual(getDomNode(handlerHiding.args[2][0].rootItem), $rootItem[0], '2nd level hide');
+        assert.strictEqual(handlerHiding.args[2][0].itemData.id, 12);
+        assert.strictEqual(getDomNode(handlerHiding.args[2][0].submenuContainer), nestedSubmenu2);
+        assert.strictEqual(getDomNode(handlerHidden.args[1][0].rootItem), $rootItem[0]);
+        assert.strictEqual(handlerHidden.args[1][0].itemData.id, 12);
+        assert.strictEqual(getDomNode(handlerHidden.args[1][0].submenuContainer), nestedSubmenu2);
+    });
```

Comment:
[code style]
Please refactor it. It's totally unreadable now.
1. Do not use private API. To already have `rootSubmenu` value found in public way, but below you do it one more time using private API.
2. Reduce copypaste. Now you get same elements several times. You can also extract parameters checking to a method.
3. Why do we need to check parameters both for submenu1 and submenu2? Seems like the test cannot be red for second menu of it's greed for the first one.
4. Check the whole itemData equality, not only id. It will not only improve coverage but make the test more intuitive and clear: `handlerShowing.args[0][0].itemData === options.items[0]`
5. Add meaningful assert comments and update the test name.

Actually I also prefer splitting this test. It's okay to leave the single test if it's readable enough tho.